shared/src/main/scala/coop/rchain/shared/AtomicSyncVar.scala

```diff
@@ -0,0 +1,30 @@
+package coop.rchain.shared
+
+import scala.concurrent.SyncVar
+import scala.util.{Failure, Success, Try}
+
+//A SyncVar which enforces the "consume/replace" model of
+//atomic updates (same as Rholang's for/! pattern with a
+//linear send).
+class AtomicSyncVar[A](init: A) {
+  private[this] val underlying: SyncVar[A] = new SyncVar()
+  underlying.put(init)
+
+  def get: A = underlying.get
+
+  //Use case: capture an intermediate result in the process of updating the var.
+  def mapAndUpdate[B](f: (A) => B, g: (B) => A): Either[Throwable, B] = {
```

Comment:
I am just trying to write down possible threads of execution here to understand what is (or can be) going on here:
1) if `f` doesn't throw any exception then what we get back is what was inside with `f ` applied ( `underlying.get andThen f`) and we put back what was inside with `f andThen g` applied
2) if either `f` or `g` throws then value inside doesn't change  and we return an exception
is that right?

What worries me is that the thing that is put back after update will go through two modifying functions (`f` and `g`) which is kind of unintuitive. Why not expose simple `update` method that will just `take`, modify and `put` value back and push all other modifications of the value to the client ?

Reply:
I do expose exactly such a method here https://github.com/rchain/rchain/pull/725/files#diff-9e5ff89d6d9c0f94bc7f6a2f503e3e53R28
As I say in the comment, the use case for this method over `update` is to capture an intermediate result during the computation of the updated value. For example, if `h: A => A = g compose f`, and I call `update(h)`, then the final value of the var would be the same as `mapAndUpdate(f, g)`, but value which is returned would be different.  Of course, I could do something like this:
```scala
val curr = atomicVar.get
val interm = f(curr)
val updated = atomicVar.update(h)
...
```
But I think this is worse because (a) I needed to do the work of `f` twice and (b) it's possible for another thread to modify the var between the `get` and `update`, so the final result may not even be what I expect.
The `mapAndUpdate` method may not be terribly intuitive, but it is precisely what I need for situations like this: https://github.com/rchain/rchain/pull/725/files#diff-847a2b93af9873d8612f90d45a43b6c5R156

Reply:
Ok, fine by me then. But you have to be careful and remember to not expose any `put` method on its own because that will require additional locks to handle concurrent access.